qiskit/circuit/library/evolution/hamiltonian_gate.py

```diff
@@ -0,0 +1,136 @@
+        # Check input is N-qubit matrix
+        input_dim, output_dim = data.shape
+        num_qubits = int(numpy.log2(input_dim))
+        if input_dim != output_dim or 2**num_qubits != input_dim:
+            raise ValueError("Input matrix is not an N-qubit operator.")
+
+        # Store instruction params
+        super().__init__("hamiltonian", num_qubits, [data, time], label=label)
+
+    def __eq__(self, other):
+        if not isinstance(other, HamiltonianGate):
+            return False
+        if self.label != other.label:
+            return False
+        operators_eq = matrix_equal(self.params[0], other.params[0], ignore_phase=False)
+        times_eq = self.params[1] == other.params[1]
+        return operators_eq and times_eq
+
+    def __array__(self, dtype=None):
+        """Return matrix for the unitary."""
+        # pylint: disable=unused-argument
+        import scipy.linalg
+
+        try:
+            return scipy.linalg.expm(-1j * self.params[0] * float(self.params[1]))
+        except TypeError as ex:
+            raise TypeError(
+                "Unable to generate Unitary matrix for "
+                "unbound t parameter {}".format(self.params[1])
+            ) from ex
+
+    def inverse(self):
+        """Return the adjoint of the unitary."""
+        return self.adjoint()
+
+    def conjugate(self):
+        """Return the conjugate of the Hamiltonian."""
+        return HamiltonianGate(numpy.conj(self.params[0]), -self.params[1])
+
+    def adjoint(self):
+        """Return the adjoint of the unitary."""
+        return HamiltonianGate(self.params[0], -self.params[1])
+
+    def transpose(self):
+        """Return the transpose of the Hamiltonian."""
+        return HamiltonianGate(numpy.transpose(self.params[0]), self.params[1])
+
+    def _define(self):
+        """Calculate a subcircuit that implements this unitary."""
+        qc = QuantumCircuit(self.num_qubits, name=self.name)
+        qc._append(UnitaryGate(self.to_matrix()), qc.qubits, [])
+        self.definition = qc
+
+    def qasm(self):
+        """Raise an error, as QASM is not defined for the HamiltonianGate."""
+        raise ValueError("HamiltonianGate has no QASM definition.")
+
+    def validate_parameter(self, parameter):
+        """Hamiltonian parameter has to be an ndarray, operator or float."""
+        if isinstance(parameter, (float, int, numpy.ndarray)):
+            return parameter
+        elif isinstance(parameter, ParameterExpression) and len(parameter.parameters) == 0:
+            return float(parameter)
+        else:
+            raise CircuitError(f"invalid param type {type(parameter)} for gate {self.name}")
+
+
+def hamiltonian(self, operator, time, qubits, label=None):
+    """Apply hamiltonian evolution to qubits."""
+    if not isinstance(qubits, list):
+        qubits = [qubits]
+
+    return self.append(HamiltonianGate(data=operator, time=time, label=label), qubits, [])
+
+
+QuantumCircuit.hamiltonian = hamiltonian
```

Comment:
The docs artifact from the CI build seems to be too long ago now for me to inspect it, but will this method be properly documented when injected into `QuantumCircuit` like this?

Reply:
Yes 🙂 you can check for some other methods monkey patched in the same fashion, such as `QuantumCircuit.unitary`.